Fix Numpy and Scipy deprecation warnings

yanbomb opened the pull request:
Fix Numpy and Scipy deprecation warnings.

Changes:
- numpy.product() -> numpy.prod()
- array to float conversion for functions used by scipy.integrate.quad/quadn and scipy.optimize.brentq

Setup:
- Python 3.12.2
- Numpy 1.26.4
- Scipy 1.12.0

Verification:

 I ran pytest and got 573 passed tests; 0 failues; 43 skipped (also ran these tests; all passed); 94 warnings. The warning are related to: a) Silverman and Scott methods with multiples weights (92 warnings), b) datetime deprecation (does not seem to be related to KDEpy)

Note:
There is another instance of brentq in bw_selection.py that did not emit warnings during testing. Code was left as is, but more testing could eventually reveal additional deprecation warnings. The easy fix would be to wrap the function using a lambda function:
```
ff=lambda x:function(x)[0]
x, res = brentq(ff, 0, tol, args=args, full_output=True, disp=False)
```

tommyod commented:
Thanks a lot! Changes look good. Could you run the linter too, so the CI pipeline passes? Probably just some whitespace

yanbomb commented:
I made some changes, but this is all new to me so I hope that I correctly addressed your concerns. While introducing myself to pylint and PEP-8, I noticed that many locations may not comply with PEP-8 (mostly regarding naming conventions and unused variables).

P.S. The major issue was that the usage of lambdas is not permitted. That was a surprise to me...

yanbomb commented:
Made changes that I hope will fix the formatting issues and ran the following checks which resulted in no error:

- python -m black KDEpy -l 120 --check
- python -m flake8 --show-source --ignore=F811,W293,W391,W292,W291,W504,W503,E231 --max-line-length=120 --exclude="*examples.py,testing.py,*kde.py" KDEpy
- pytest KDEpy --doctest-modules --capture=sys

P.S. Sorry for the iterative process for something that should be trivial... I am learning something new here with flake8 and black.

whitews commented:
Hi @yanbomb 

Looks like black is getting caught on the lack of empty lines before & after the nested functions. She's a stickler for the rules ;)

An example from the diff:
```
@@ -54,12 +54,14 @@
 
         if function.finite_support:
             a, b = -function.support, function.support
         else:
             a, b = -5 * function.var, 5 * function.var
+
         def function_float(x):
             return function(x)[0]
+
         integral, _ = quad(function_float, a=a, b=b)
         assert np.isclose(integral, 1)
```

Dunno why the Windows tests are passing, but that should help you get the PEP8 stuff cleared up.

yanbomb commented:
I finally got the formatting correct! Thank you @tommyod for your patience and @whitews for the hint.
P.S. That was my first PR...

whitews commented:
Nice! I triggered the tests and it all looks good, I'll leave for Tommy to merge.

yanbomb commented:
Should I click the "close with comment" button to close this PR?

whitews commented:
Nope, you're all good. Tommy will have the final say & merge. He may want to bump the version for a new release to PyPI.

tommyod commented:
Great job on your first PR @yanbomb . Thanks for the help @whitews !

I think you're all set. I'll merge later today or tomorrow, then I will bump versions at the same time and.
